components/opc-ua/src/Microsoft.Azure.IIoT.OpcUa.Protocol/src/Services/SubscriptionServices.cs

```diff
@@ -1249,22 +1272,32 @@ private class MonitoredItemWrapper {
                 }
                 while (superType != null);
 
-                var propertyNames = new List<QualifiedName>();
+                var fieldNames = new List<QualifiedName>();
                 foreach (var node in nodes) {
-                    foreach (var reference in node.ReferenceTable) {
-                        if (reference.ReferenceTypeId == ReferenceTypeIds.HasProperty) {
-                            propertyNames.Add(nodeCache.FetchNode(reference.TargetId).BrowseName);
-                        }
-                    }
+                    ParseFields(nodeCache, fieldNames, node);
                 }
+                fieldNames = fieldNames
+                    .OrderBy(x => x.Name).ToList();
 
                 var eventFilter = new EventFilter();
-                foreach (var propertyName in propertyNames) {
+                // Let's add ConditionId manually first if event is derived from ConditionType
+                if (nodes.Any(x => x.NodeId == ObjectTypeIds.ConditionType)) {
+                    eventFilter.SelectClauses.Add(new SimpleAttributeOperand() {
+                        BrowsePath = new QualifiedNameCollection(),
+                        TypeDefinitionId = ObjectTypeIds.ConditionType,
+                        AttributeId = Attributes.NodeId
+                    });
+                }
+
+                foreach (var fieldName in fieldNames) {
```

Comment:
There is actually an implicit conversion of the QualifiedNameCollection that would allow you to make this loop really compact. Something like:

```
foreach (var fieldName in fieldNames) {
    var selectClause = new SimpleAttributeOperand() {
        TypeDefinitionId = ObjectTypeIds.BaseEventType,
        AttributeId = Attributes.Value,
        BrowsePath = fieldName.Name
            .Split('|')
            .Select(x => new QualifiedName(x, fieldName.NamespaceIndex))
            .ToArray()
    };
    eventFilter.SelectClauses.Add(selectClause);
}
```

Reply:
Nice, fixed.